subsys/nsettings/settings_fcb.c

```diff
@@ -0,0 +1,550 @@
+/* Copyright (c) 2019 Laczen
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "settings_fcb.h"
+#include "settings_priv.h"
+
+struct settings_fcb_read_fn_arg {
+	struct fcb_entry_ctx *entry_ctx;
+	off_t off;
+};
+
+static int settings_fcb_load(struct settings_store *cs);
+static int settings_fcb_save(struct settings_store *cs, const char *name,
+			     const char *value, size_t val_len);
+
+static struct settings_store_itf settings_fcb_itf = {
+	.csi_load = settings_fcb_load,
+	.csi_save = settings_fcb_save,
+};
+
+int settings_fcb_dst(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_dst_register(&cf->cf_store);
+
+	return 0;
+}
+
+int settings_fcb_src(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_src_register(&cf->cf_store);
+
+	return 0;
+}
+
+static int settings_fcb_read(struct fcb_entry_ctx *entry_ctx, off_t off,
+			     char *buf, size_t *len)
+{
+	if (off >= entry_ctx->loc.fe_data_len) {
+		return -EINVAL;
+	}
+
+	if ((off + *len) > entry_ctx->loc.fe_data_len) {
+		*len = entry_ctx->loc.fe_data_len - off;
+	}
+
+	return flash_area_read(entry_ctx->fap,
```

Comment:
flash read operation should be aligned. Unaligned access is not enforced by `read` API.

Reply:
The routine is a copy of what was in the settings module.

Reply:
Not exactly - fcb's `read_handler` was called by routine chains which took care about aligned operations. This is implemented in `settings_line_raw_read_until`.

Reply:
You are right, the old settings did handle alignment, I will need to change the read.